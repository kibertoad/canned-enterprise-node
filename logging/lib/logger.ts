import Pino, { Logger as PinoLogger, LoggerOptions, stdTimeFunctions } from 'pino'
import config from './config'
import { extendMeta } from './loggerUtils'
import { Context } from './types'

interface ValueMap {
  [key: string]: any
}

export const defaultPrettyPrintOptions = Object.freeze({
  translateTime: true
})

export const defaultOptions: LoggerOptions = Object.freeze({
  timestamp: stdTimeFunctions.epochTime,
  level: config.LOG_LEVEL,
  useLevelLabels: true,
  messageKey: 'message',
  prettyPrint: config.USE_PRETTY_LOGS || config.IS_DEVELOPMENT ? defaultPrettyPrintOptions : false,
  base: null
})

export class Logger {
  private pinoLogger: PinoLogger

  constructor(optionOverrides: Partial<LoggerOptions> = {}, loggerInstance?: PinoLogger) {
    this.pinoLogger =
      loggerInstance ||
      Pino({
        ...defaultOptions,
        ...optionOverrides
      })
  }

  info(ctx: Context | null, message: string, meta: object = {}): void {
    this.pinoLogger.info(extendMeta(ctx, meta), message)
  }

  debug(ctx: Context | null, message: string, meta: object = {}): void {
    this.pinoLogger.debug(extendMeta(ctx, meta), message)
  }

  warn(ctx: Context | null, message: string, meta: object = {}): void {
    const effectiveMeta = {
      ...extendMeta(ctx, meta)
    }
    this.pinoLogger.warn(effectiveMeta, message)
  }

  error(ctx: Context | null, message: string, err?: Error, meta: object = {}): void {
    const errLog = err ? err : {}
    const effectiveMeta = {
      ...extendMeta(ctx, meta),
      errLog
    }
    this.pinoLogger.error(effectiveMeta, message)
  }

  childLogger(baseValues: ValueMap): Logger {
    return new Logger(undefined, this.pinoLogger.child(baseValues))
  }
}

export const defaultLogger = new Logger()
