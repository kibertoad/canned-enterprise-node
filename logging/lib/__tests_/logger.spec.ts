import Mockdate from 'mockdate'
import { defaultPrettyPrintOptions, Logger } from '../logger'

describe('logger', () => {
  const consoleSpy = jest.spyOn(process.stdout, 'write')
  let logger: Logger

  async function resetLogger(level: string | undefined) {
    process.env.LOG_LEVEL = level
    jest.resetModules()
    logger = (await import('../logger')).defaultLogger
  }

  beforeAll(async () => {
    const mockedDate = new Date('2011-10-05T14:48:00.000Z')
    Mockdate.set(mockedDate)
    await resetLogger('debug')
  })

  afterEach(async () => {
    consoleSpy.mockReset()
    await resetLogger('debug')
  })

  afterAll(() => {
    Mockdate.reset()
  })

  it('does not log DEBUG info when logger is set to INFO level', async () => {
    await resetLogger('info')

    logger.debug({ requestId: 'dummy' }, 'Dummy text', {})
    expect(consoleSpy).not.toHaveBeenCalled()
  })

  it('logs INFO level entry', () => {
    logger.info({ requestId: 'dummy' }, 'Dummy text', {})
    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  // Skip it until we figure out why colour codes are omitted in CI
  it.skip('logs pretty INFO level entry', () => {
    const prettyLogger = new Logger({
      prettyPrint: defaultPrettyPrintOptions,
      level: 'debug',
      name: 'dummy'
    })
    prettyLogger.info({ requestId: 'dummy' }, 'Dummy text', {})
    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('logs DEBUG level entry with context data', () => {
    logger.debug(
      {
        requestId: 'dummyId'
      },
      'Request-header validation passed.',
      {
        originalHost: 'localhost'
      }
    )

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('Logs ERROR level entry with error data', () => {
    const e: any = new Error('Dummy error message')
    e.additionalFields = {
      productId: 'dummyId'
    }
    logger.error(
      {
        requestId: 'dummyId'
      },
      e.message,
      undefined,
      e.additionalFields
    )

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('Logs ERROR level entry with error', () => {
    const e: any = new Error('Dummy error message')
    e.additionalFields = {
      productId: 'dummyId'
    }
    e.stack = 'Dummy stack'
    logger.error(
      {
        requestId: 'dummyId'
      },
      e.message,
      e,
      e.additionalFields
    )

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('Logs WARN level entry with error data', () => {
    const e: any = new Error('Dummy error message')
    e.additionalFields = {
      productId: 'dummyId'
    }
    logger.warn(
      {
        requestId: 'dummyId'
      },
      e.message,
      e.additionalFields
    )

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('Supports setting requestId in childLogger', () => {
    const childLogger = logger.childLogger({ requestId: 'abc' })

    childLogger.info(null, 'dummy')

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })

  it('Supports overriding childLogger base values', () => {
    const childLogger = logger.childLogger({ serviceId: 'abc' })

    childLogger.info(null, 'dummy2', { serviceId: 'xyz' })

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
  })
})
