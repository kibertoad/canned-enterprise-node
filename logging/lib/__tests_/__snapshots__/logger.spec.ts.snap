// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger Logs ERROR level entry with error 1`] = `
"[2011-10-05 14:48:00.000 +0000] [31mERROR[39m: [36mDummy error message[39m
    productId: \\"dummyId\\"
    requestId: \\"dummyId\\"
    errLog: {
      \\"additionalFields\\": {
        \\"productId\\": \\"dummyId\\"
      }
    }
"
`;

exports[`logger Logs ERROR level entry with error data 1`] = `
"[2011-10-05 14:48:00.000 +0000] [31mERROR[39m: [36mDummy error message[39m
    productId: \\"dummyId\\"
    requestId: \\"dummyId\\"
    errLog: {}
"
`;

exports[`logger Logs WARN level entry with error data 1`] = `
"[2011-10-05 14:48:00.000 +0000] [33mWARN [39m: [36mDummy error message[39m
    productId: \\"dummyId\\"
    requestId: \\"dummyId\\"
"
`;

exports[`logger Supports overriding childLogger base values 1`] = `
"[2011-10-05 14:48:00.000 +0000] [32mINFO [39m: [36mdummy2[39m
    serviceId: \\"xyz\\"
"
`;

exports[`logger Supports setting requestId in childLogger 1`] = `
"[2011-10-05 14:48:00.000 +0000] [32mINFO [39m: [36mdummy[39m
    requestId: \\"abc\\"
"
`;

exports[`logger logs DEBUG level entry with context data 1`] = `
"[2011-10-05 14:48:00.000 +0000] [94mDEBUG[39m: [36mRequest-header validation passed.[39m
    originalHost: \\"localhost\\"
    requestId: \\"dummyId\\"
"
`;

exports[`logger logs INFO level entry 1`] = `
"[2011-10-05 14:48:00.000 +0000] [32mINFO [39m: [36mDummy text[39m
    requestId: \\"dummy\\"
"
`;
